third_party/pyth/price-service/src/ws.ts

```diff
@@ -0,0 +1,211 @@
+
+      client.send(JSON.stringify(priceUpdate));
+    }
+  }
+
+  clientClose(ws: WebSocket) {
+    for (let clients of this.priceFeedClients.values()) {
+      if (clients.has(ws)) {
+        clients.delete(ws);
+      }
+    }
+
+    this.aliveClients.delete(ws);
+    this.wsId.delete(ws);
+  }
+
+  handleMessage(ws: WebSocket, data: RawData) {
+    try {
+      let jsonData = JSON.parse(data.toString());
+      let validationResult = ClientMessageSchema.validate(jsonData);
+      if (validationResult.error !== undefined) {
+        throw validationResult.error;
+      }
+
+      let message = jsonData as ClientMessage;
+
+      message.ids = message.ids.map((id) => {
+        if (id.startsWith("0x")) {
+          return id.substring(2);
+        }
+        return id;
+      });
+
+      const availableIds = this.priceFeedVaaInfo.getPriceIds();
+      let notFoundIds = message.ids.filter((id) => !availableIds.has(id));
+
+      if (notFoundIds.length > 0) {
+        throw new Error(`Price Feeds with ids ${notFoundIds.join(', ')} not found`)
+      }
+
+      if (message.type == "subscribe") {
+        message.ids.forEach( id => this.addPriceFeedClient(ws, id) );
+      } else {
+        message.ids.forEach( id => this.delPriceFeedClient(ws, id) );
+      }
+    } catch (e: any) {
+      let response: ServerResponse = {
+        type: "response",
+        status: "error",
+        error: e.message
+      };
+
+      logger.info(`Invalid request ${data.toString()} from client ${this.wsId.get(ws)}`);
+      this.promClient?.addWebSocketInteraction("client_message", "err");
+
+      ws.send(JSON.stringify(response));
+      return;
+    }
+
+    logger.info(`Successful request ${data.toString()} from client ${this.wsId.get(ws)}`);
+    this.promClient?.addWebSocketInteraction("client_message", "ok");
+
+    let response: ServerResponse = {
+      type: "response",
+      status: "success",
+    }
+    
+    ws.send(JSON.stringify(response));
+  }
+
+  run(): [WebSocketServer, http.Server] {
+    const app = express();
+    const server = http.createServer(app);
+
+    const wss = new WebSocketServer({ server });
+
+    wss.on('connection', (ws: WebSocket, request: http.IncomingMessage) => {
+      logger.info(`Incoming ws connection from ${request.socket.remoteAddress}, assigned id: ${this.wsCounter}`)
+
+      this.wsId.set(ws, this.wsCounter);
+      this.wsCounter += 1;
+
+      ws.on("message", (data: RawData) => this.handleMessage(ws, data));
+      
+      this.aliveClients.add(ws);
+
+      ws.on("pong", (_data) => {
+        this.aliveClients.add(ws);
+      });
+
+      ws.on("close", (_code: number, _reason: Buffer) => {
+        logger.info(`client ${this.wsId.get(ws)} closed the connection.`);
+        this.promClient?.addWebSocketInteraction("close", "ok");
+
+        this.clientClose(ws);
+      });
+
+      this.promClient?.addWebSocketInteraction("connection", "ok");
+    });
+
+    const pingInterval = setInterval( () => {
```

Comment:
Overall looks great, a quick format with `prettier` or similar run over the files to fix up little spacing issues is the only nitpick.